File: 51. N-Queens.cpp
// Recursion Backtracing ]
//TC - O(N! * N) SC - O(N^2)
class Solution {
public:
    bool isSafe(int i, int j, int n, vector<string> &board){
        for(int row = 0; row<i; row++){
            if(board[row][j] == 'Q') return false;
        }

        int row = i;
        int col = j;

        while(row >= 0 && col >= 0){
            if(board[row][col] == 'Q')    return false;

            row--;
            col--;
        }

        row = i;
        col = j;

        while(row >= 0 && col < n){
            if(board[row][col] == 'Q') return false;

            row--;
            col++;
        }
        return true;
    }

    void solve(int row,const int &n, vector<string> &board, vector<vector<string>> &ans){
        if(row == n){
            ans.push_back(board);
            return;
        }

        for(int col = 0; col<n; col++){
            if(isSafe(row, col, n, board)){
                board[row][col] = 'Q';
                solve(row+1, n, board, ans);
                board[row][col] = '.';
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        string s;
        for(int i=0;i<n; i++){
            s.push_back('.');
        }
        vector<string> board(n, s);
        vector<vector<string>> ans;
        
        solve(0, n, board, ans);
        return ans;
    }    
};

// Optimal approach
//SC - O(N)
class Solution {
public:
    void solve(int row,const int &n, vector<int> &cols, vector<int> &leftd, vector<int> &rightd, vector<string> &board, vector<vector<string>> &ans){
        if(row == n){
            ans.push_back(board);
            return;
        }

        for(int col = 0; col<n; col++){
            if(cols[col] == 0 && leftd[row+col] == 0 && rightd[(n-1) + (col-row)] == 0){
                cols[col] = 1;
                leftd[row+col] = 1;
                rightd[(n-1) + (col-row)] = 1;
                board[row][col] = 'Q';
                solve(row+1, n, cols, leftd, rightd, board, ans);
                board[row][col] = '.';
                cols[col] = 0;
                leftd[row+col] = 0;
                rightd[(n-1) + (col-row)] = 0;
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        string s(n, '.');
        vector<string> board(n, s);
        vector<vector<string>> ans;
        
        vector<int> cols(n, 0);
        vector<int> leftd(2*n - 1, 0);
        vector<int> rightd(2*n - 1, 0);

        solve(0, n, cols, leftd, rightd, board, ans);
        return ans;
    }    
};

File: 51. N-Queens.java
//Optimal Soln
class Solution {

    public void saveBoard(char[][] board, List<List<String>> ans){
        String row = "";
        List<String> newBoard = new ArrayList<>();

        for(int i=0; i<board.length; i++){
            row = "";
            for(int j=0; j<board[0].length; j++){
                if(board[i][j] == 'Q')
                    row += 'Q';
                else
                    row += '.';
            }
            newBoard.add(row);
        }

        ans.add(newBoard);
    }

    public void solve(int row, int n, int[] rows, int[] leftd, int[] rightd, char[][] board, List<List<String>> ans){
        if(row == n){
            saveBoard(board, ans);
            return;
        }

        for(int col = 0; col<n ; col++){
            if(rows[col] == 0 && leftd[row+col] == 0 && rightd[n-1 + col-row] == 0){
                board[row][col] = 'Q';
                rows[col] = 1;
                leftd[row+col] = 1;
                rightd[n-1 + col-row] = 1;
                solve(row+1, n, rows, leftd, rightd, board, ans);
                board[row][col] = '.';
                rows[col] = 0;
                leftd[row+col] = 0;
                rightd[n-1 + col-row] = 0;
            }
        }
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<>();
        char[][] board = new char[n][n];

        int[] rows = new int[n];
        int[] leftd = new int[2*n - 1];
        int[] rightd = new int[2*n - 1];
 
        solve(0, n, rows, leftd, rightd, board, ans);
        return ans;
    }
}


// Recursive Soln
class Solution {

    public Boolean isSafe(int row, int col, char[][] board){
        for(int i = 0; i<row ; i++){
            if(board[i][col] == 'Q'){
                return false;
            }
        }

        int r = row;
        int c = col;

        while(r>=0 && c>=0){
            if(board[r][c] == 'Q'){
                return false;
            }
            r--;
            c--;
        }

        r = row;
        c = col;

        while(r>=0 && c<board.length){
            if(board[r][c] == 'Q'){
                return false;
            }
            r--;
            c++;
        }

        return true;
    }

    public void saveBoard(char[][] board, List<List<String>> ans){
        String row = "";
        List<String> newBoard = new ArrayList<>();

        for(int i=0; i<board.length; i++){
            row = "";
            for(int j=0; j<board[0].length; j++){
                if(board[i][j] == 'Q')
                    row += 'Q';
                else
                    row += '.';
            }
            newBoard.add(row);
        }

        ans.add(newBoard);
    }

    public void solve(int row, int n, char[][] board, List<List<String>> ans){
        if(row == n){
            saveBoard(board, ans);
            return;
        }

        for(int col = 0; col<n ; col++){
            if(isSafe(row, col, board)){
                board[row][col] = 'Q';
                solve(row+1, n, board, ans);
                board[row][col] = '.';
            }
        }
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<>();
        char[][] board = new char[n][n];

        solve(0, n, board, ans);
        return ans;
    }
}
